Annotations: Allow at symbol in URLs

pando85 opened the pull request:
The same as #11739, we need to allow those kinds of configurations in case your group names contain your domain separated by an at symbol. 

## What this PR does / why we need it:
To allow group names that contain the domain. 

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] CVE Report (Scanner found CVE and adding report)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation only

## Which issue/s this PR fixes
<!--
(optional, in `fixes #<issue number>` format, will close that issue when PR gets merged):

fixes #
-->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/kubernetes/ingress-nginx/blob/main/CONTRIBUTING.md) guide
- [x] I have added unit and/or e2e tests to cover my changes.
- [x] All new and existing tests passed.

Gacko commented on internal/ingress/annotations/parser/validators_test.go:
```diff
@@ -60,6 +60,11 @@ func TestValidateArrayOfServerName(t *testing.T) {
 			wantErr: false,
 		},
+		{
+			name:    "should allow comma separated query params",
```
```suggestion
			name:    "should allow @ in query params",
```

Gacko commented:
Please review your change, `CI / test-go` is failing for it.

pando85 commented:
> Please review your change, `CI / test-go` is failing for it.

Do you think it is related with my change? I ran it in local and I have the same result with this commit and with the previous one.

pando85 commented:
I just rebase main again to see if that fixes the problem.

pando85 commented:
I found the error. I added the test in the same place as #11739 but that was not checking the [correct parser](https://github.com/kubernetes/ingress-nginx/blob/5b142ed7c44327a949037833531b7cf11057a1ad/internal/ingress/annotations/authreq/main.go#L60).

I didn't find any test for `URLWithNginxVariableRegex` but I checked that the current change should fix the problem that I described.

I will amend the commit removing the wrong lines that I added to the unit tests.

Gacko commented:
Adding a test is not about proving your change works, it's about ensuring it keeps working with future changes. So I'd ask you to add a test.

pando85 commented:
> Adding a test is not about proving your change works, it's about ensuring it keeps working with future changes. So I'd ask you to add a test.

Oh yes, I agree with that. I probably didn't communicate it well. Let's try again.

Could you point me to the current tests related to those validators? I cannot find them.

Also, PR #11882, which is pretty similar to this one, added one useless test. It isn't testing the behavior, indeed you can revert the change in `validators.go` and the tests still pass with that new test added.

strongjz commented:
I did test this. The Go unit test still passed with and without my change, you #11882. That is unfortunate. 

We need a generic e2e to test this behavior without relying on any one set of annotations.
